File: control/src/main/java/spp/probe/control/LiveInstrumentRemote.java
package spp.probe.control;

import io.vertx.core.AbstractVerticle;
import io.vertx.core.json.Json;
import io.vertx.core.json.JsonObject;
import io.vertx.ext.bridge.BridgeEventType;
import io.vertx.ext.eventbus.bridge.tcp.impl.protocol.FrameHelper;
import org.apache.skywalking.apm.agent.core.context.util.ThrowableTransformer;
import org.apache.skywalking.apm.agent.core.plugin.WitnessFinder;
import spp.probe.SourceProbe;
import spp.protocol.instrument.LiveInstrument;
import spp.protocol.instrument.breakpoint.LiveBreakpoint;
import spp.protocol.instrument.log.LiveLog;
import spp.protocol.instrument.meter.LiveMeter;
import spp.protocol.platform.PlatformAddress;
import spp.protocol.probe.command.LiveInstrumentCommand;

import java.lang.instrument.Instrumentation;
import java.lang.reflect.Field;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.HashMap;
import java.util.Map;
import java.util.function.BiConsumer;

import static spp.probe.SourceProbe.PROBE_ID;
import static spp.probe.SourceProbe.instrumentation;
import static spp.protocol.probe.ProbeAddress.*;

public class LiveInstrumentRemote extends AbstractVerticle {

    private static final BiConsumer<String, String> EVENT_CONSUMER = (address, json) -> FrameHelper.sendFrame(
            BridgeEventType.PUBLISH.name().toLowerCase(), address, new JsonObject(json), SourceProbe.tcpSocket
    );

    private Method applyInstrument;
    private Method removeInstrument;
    private static Method putBreakpoint;
    private static Method putLog;
    private static Method putMeter;
    private static Method isInstrumentEnabled;
    private static Method putLocalVariable;
    private static Method putField;
    private static Method putStaticField;
    private static Method isHit;

    @Override
    public void start() {
        try {
            ClassLoader agentClassLoader = (ClassLoader) Class.forName("org.apache.skywalking.apm.agent.core.plugin.loader.AgentClassLoader")
                    .getMethod("getDefault").invoke(null);
            Class serviceClass = Class.forName(
                    "spp.probe.services.instrument.LiveInstrumentService", false, agentClassLoader);
            Field poolMapField = WitnessFinder.class.getDeclaredField("poolMap");
            poolMapField.setAccessible(true);
            serviceClass.getMethod("setPoolMap", Map.class).invoke(null, poolMapField.get(WitnessFinder.INSTANCE));
            serviceClass.getMethod("setInstrumentEventConsumer", BiConsumer.class).invoke(null, EVENT_CONSUMER);
            serviceClass.getMethod("setInstrumentation", Instrumentation.class).invoke(null, instrumentation);

            applyInstrument = serviceClass.getMethod("applyInstrument", LiveInstrument.class);
            removeInstrument = serviceClass.getMethod("removeInstrument",
                    String.class, int.class, String.class);
            isInstrumentEnabled = serviceClass.getMethod("isInstrumentEnabled", String.class);
            isHit = serviceClass.getMethod("isHit", String.class);

            Class contextClass = Class.forName(
                    "spp.probe.services.common.ContextReceiver", false, agentClassLoader);
            putLocalVariable = contextClass.getMethod("putLocalVariable",
                    String.class, String.class, Object.class);
            putField = contextClass.getMethod("putField",
                    String.class, String.class, Object.class);
            putStaticField = contextClass.getMethod("putStaticField",
                    String.class, String.class, Object.class);
            putBreakpoint = contextClass.getMethod("putBreakpoint",
                    String.class, String.class, int.class, Throwable.class);
            putLog = contextClass.getMethod("putLog",
                    String.class, String.class, String[].class);
            putMeter = contextClass.getMethod("putMeter", String.class);
        } catch (Throwable e) {
            e.printStackTrace();
            throw new RuntimeException(e);
        }

        vertx.eventBus().<JsonObject>localConsumer("local." + LIVE_BREAKPOINT_REMOTE.getAddress() + ":" + PROBE_ID).handler(it -> {
            try {
                LiveInstrumentCommand command = Json.decodeValue(it.body().toString(), LiveInstrumentCommand.class);
                switch (command.getCommandType()) {
                    case ADD_LIVE_INSTRUMENT:
                        addBreakpoint(command);
                        break;
                    case REMOVE_LIVE_INSTRUMENT:
                        removeInstrument(command);
                        break;
                }
            } catch (InvocationTargetException ex) {
                Map<String, Object> map = new HashMap<>();
                map.put("command", it.body().toString());
                map.put("occurredAt", System.currentTimeMillis());
                if (ex.getCause() != null) {
                    map.put("cause", ThrowableTransformer.INSTANCE.convert2String(ex.getCause(), 4000));
                } else {
                    map.put("cause", ThrowableTransformer.INSTANCE.convert2String(ex.getTargetException(), 4000));
                }

                FrameHelper.sendFrame(
                        BridgeEventType.PUBLISH.name().toLowerCase(),
                        PlatformAddress.LIVE_BREAKPOINT_REMOVED.getAddress(),
                        JsonObject.mapFrom(map), SourceProbe.tcpSocket
                );
            } catch (Throwable ex) {
                Map<String, Object> map = new HashMap<>();
                map.put("command", it.body().toString());
                map.put("occurredAt", System.currentTimeMillis());
                map.put("cause", ThrowableTransformer.INSTANCE.convert2String(ex, 4000));

                FrameHelper.sendFrame(
                        BridgeEventType.PUBLISH.name().toLowerCase(),
                        PlatformAddress.LIVE_BREAKPOINT_REMOVED.getAddress(),
                        JsonObject.mapFrom(map), SourceProbe.tcpSocket
                );
            }
        });
        vertx.eventBus().<JsonObject>localConsumer("local." + LIVE_LOG_REMOTE.getAddress() + ":" + PROBE_ID).handler(it -> {
            try {
                LiveInstrumentCommand command = Json.decodeValue(it.body().toString(), LiveInstrumentCommand.class);
                switch (command.getCommandType()) {
                    case ADD_LIVE_INSTRUMENT:
                        addLog(command);
                        break;
                    case REMOVE_LIVE_INSTRUMENT:
                        removeInstrument(command);
                        break;
                }
            } catch (InvocationTargetException ex) {
                Map<String, Object> map = new HashMap<>();
                map.put("command", it.body().toString());
                map.put("occurredAt", System.currentTimeMillis());
                if (ex.getCause() != null) {
                    map.put("cause", ThrowableTransformer.INSTANCE.convert2String(ex.getCause(), 4000));
                } else {
                    map.put("cause", ThrowableTransformer.INSTANCE.convert2String(ex.getTargetException(), 4000));
                }

                FrameHelper.sendFrame(
                        BridgeEventType.PUBLISH.name().toLowerCase(),
                        PlatformAddress.LIVE_LOG_REMOVED.getAddress(),
                        JsonObject.mapFrom(map), SourceProbe.tcpSocket
                );
            } catch (Throwable ex) {
                Map<String, Object> map = new HashMap<>();
                map.put("command", it.body().toString());
                map.put("occurredAt", System.currentTimeMillis());
                map.put("cause", ThrowableTransformer.INSTANCE.convert2String(ex, 4000));

                FrameHelper.sendFrame(
                        BridgeEventType.PUBLISH.name().toLowerCase(),
                        PlatformAddress.LIVE_LOG_REMOVED.getAddress(),
                        JsonObject.mapFrom(map), SourceProbe.tcpSocket
                );
            }
        });
        vertx.eventBus().<JsonObject>localConsumer("local." + LIVE_METER_REMOTE.getAddress() + ":" + PROBE_ID).handler(it -> {
            try {
                LiveInstrumentCommand command = Json.decodeValue(it.body().toString(), LiveInstrumentCommand.class);
                switch (command.getCommandType()) {
                    case ADD_LIVE_INSTRUMENT:
                        addMeter(command);
                        break;
                    case REMOVE_LIVE_INSTRUMENT:
                        removeInstrument(command);
                        break;
                }
            } catch (InvocationTargetException ex) {
                Map<String, Object> map = new HashMap<>();
                map.put("command", it.body().toString());
                map.put("occurredAt", System.currentTimeMillis());
                if (ex.getCause() != null) {
                    map.put("cause", ThrowableTransformer.INSTANCE.convert2String(ex.getCause(), 4000));
                } else {
                    map.put("cause", ThrowableTransformer.INSTANCE.convert2String(ex.getTargetException(), 4000));
                }

                FrameHelper.sendFrame(
                        BridgeEventType.PUBLISH.name().toLowerCase(),
                        PlatformAddress.LIVE_METER_REMOVED.getAddress(),
                        JsonObject.mapFrom(map), SourceProbe.tcpSocket
                );
            } catch (Throwable ex) {
                Map<String, Object> map = new HashMap<>();
                map.put("command", it.body().toString());
                map.put("occurredAt", System.currentTimeMillis());
                map.put("cause", ThrowableTransformer.INSTANCE.convert2String(ex, 4000));

                FrameHelper.sendFrame(
                        BridgeEventType.PUBLISH.name().toLowerCase(),
                        PlatformAddress.LIVE_METER_REMOVED.getAddress(),
                        JsonObject.mapFrom(map), SourceProbe.tcpSocket
                );
            }
        });
    }

    private void addBreakpoint(LiveInstrumentCommand command) throws Exception {
        String breakpointData = command.getContext().getLiveInstruments().get(0);
        applyInstrument.invoke(null, Json.decodeValue(breakpointData, LiveBreakpoint.class));
    }

    private void addLog(LiveInstrumentCommand command) throws Exception {
        String logData = command.getContext().getLiveInstruments().get(0);
        applyInstrument.invoke(null, Json.decodeValue(logData, LiveLog.class));
    }

    private void addMeter(LiveInstrumentCommand command) throws Exception {
        String meterData = command.getContext().getLiveInstruments().get(0);
        applyInstrument.invoke(null, Json.decodeValue(meterData, LiveMeter.class));
    }

    private void removeInstrument(LiveInstrumentCommand command) throws Exception {
        for (String breakpointData : command.getContext().getLiveInstruments()) {
            JsonObject breakpointObject = new JsonObject(breakpointData);
            String breakpointId = breakpointObject.getString("id");
            JsonObject location = breakpointObject.getJsonObject("location");
            String source = location.getString("source");
            int line = location.getInteger("line");
            removeInstrument.invoke(null, source, line, breakpointId);
        }
        for (String locationData : command.getContext().getLocations()) {
            JsonObject location = new JsonObject(locationData);
            String source = location.getString("source");
            int line = location.getInteger("line");
            removeInstrument.invoke(null, source, line, null);
        }
    }

    @SuppressWarnings("unused")
    public static boolean isInstrumentEnabled(String instrumentId) {
        try {
            return (Boolean) isInstrumentEnabled.invoke(null, instrumentId);
        } catch (Exception e) {
            e.printStackTrace();
            return false;
        }
    }

    @SuppressWarnings("unused")
    public static boolean isHit(String breakpointId) {
        try {
            return (Boolean) isHit.invoke(null, breakpointId);
        } catch (Exception e) {
            e.printStackTrace();
            return false;
        }
    }

    @SuppressWarnings("unused")
    public static void putBreakpoint(String breakpointId, String source, int line, Throwable ex) {
        try {
            putBreakpoint.invoke(null, breakpointId, source, line, ex);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @SuppressWarnings("unused")
    public static void putLog(String logId, String logFormat, String... logArguments) {
        try {
            putLog.invoke(null, logId, logFormat, logArguments);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @SuppressWarnings("unused")
    public static void putMeter(String meterId) {
        try {
            putMeter.invoke(null, meterId);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @SuppressWarnings("unused")
    public static void putLocalVariable(String breakpointId, String key, Object value) {
        try {
            putLocalVariable.invoke(null, breakpointId, key, value);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @SuppressWarnings("unused")
    public static void putField(String breakpointId, String key, Object value) {
        try {
            putField.invoke(null, breakpointId, key, value);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @SuppressWarnings("unused")
    public static void putStaticField(String breakpointId, String key, Object value) {
        try {
            putStaticField.invoke(null, breakpointId, key, value);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: control/src/main/kotlin/spp/probe/control/LiveInstrumentRemote.kt
/*
 * Source++, the open-source live coding platform.
 * Copyright (C) 2022 CodeBrig, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package spp.probe.control

import io.vertx.core.AbstractVerticle
import io.vertx.core.eventbus.Message
import io.vertx.core.json.JsonObject
import io.vertx.ext.bridge.BridgeEventType
import io.vertx.ext.eventbus.bridge.tcp.impl.protocol.FrameHelper
import org.apache.skywalking.apm.agent.core.context.util.ThrowableTransformer
import org.apache.skywalking.apm.agent.core.logging.api.LogManager
import spp.probe.ProbeConfiguration
import spp.probe.SourceProbe
import spp.protocol.instrument.*
import spp.protocol.instrument.command.CommandType
import spp.protocol.instrument.command.LiveInstrumentCommand
import spp.protocol.platform.ProbeAddress
import spp.protocol.platform.ProcessorAddress
import java.lang.instrument.Instrumentation
import java.lang.reflect.InvocationTargetException
import java.lang.reflect.Method
import java.util.*
import java.util.function.BiConsumer

@Suppress("unused")
class LiveInstrumentRemote : AbstractVerticle() {

    private var applyInstrument: Method? = null
    private var removeInstrument: Method? = null

    override fun start() {
        try {
            val agentClassLoader = Class.forName("org.apache.skywalking.apm.agent.core.plugin.loader.AgentClassLoader")
                .getMethod("getDefault").invoke(null) as ClassLoader
            val serviceClass = Class.forName(
                "spp.probe.services.instrument.LiveInstrumentService", false, agentClassLoader
            )
            serviceClass.getMethod("setInstrumentEventConsumer", BiConsumer::class.java).invoke(null, EVENT_CONSUMER)
            serviceClass.getMethod("setInstrumentation", Instrumentation::class.java)
                .invoke(null, SourceProbe.instrumentation)
            applyInstrument = serviceClass.getMethod("applyInstrument", LiveInstrument::class.java)
            removeInstrument = serviceClass.getMethod(
                "removeInstrument",
                String::class.java, Int::class.javaPrimitiveType, String::class.java
            )
            isInstrumentEnabled = serviceClass.getMethod("isInstrumentEnabled", String::class.java)
            isHit = serviceClass.getMethod("isHit", String::class.java)
            val contextClass = Class.forName(
                "spp.probe.services.common.ContextReceiver", false, agentClassLoader
            )
            putLocalVariable = contextClass.getMethod(
                "putLocalVariable",
                String::class.java, String::class.java, Any::class.java, String::class.java
            )
            putField = contextClass.getMethod(
                "putField",
                String::class.java, String::class.java, Any::class.java, String::class.java
            )
            putStaticField = contextClass.getMethod(
                "putStaticField",
                String::class.java, String::class.java, Any::class.java, String::class.java
            )
            putBreakpoint = contextClass.getMethod(
                "putBreakpoint",
                String::class.java, String::class.java, Int::class.javaPrimitiveType, Throwable::class.java
            )
            putLog = contextClass.getMethod(
                "putLog",
                String::class.java, String::class.java, Array<String>::class.java
            )
            putMeter = contextClass.getMethod("putMeter", String::class.java)
            openLocalSpan = contextClass.getMethod("openLocalSpan", String::class.java)
            closeLocalSpan = contextClass.getMethod("closeLocalSpan", String::class.java, Throwable::class.java)
        } catch (ex: Throwable) {
            log.error("Failed to initialize live instrument remote", ex)
            throw RuntimeException(ex)
        }

        vertx.eventBus() //global instrument remote
            .localConsumer<JsonObject>(ProbeAddress.LIVE_INSTRUMENT_REMOTE)
            .handler { handleInstrumentationRequest(it) }
        vertx.eventBus() //probe specific instrument remote
            .localConsumer<JsonObject>(ProbeAddress.LIVE_INSTRUMENT_REMOTE + ":" + SourceProbe.PROBE_ID)
            .handler { handleInstrumentationRequest(it) }
    }

    private fun handleInstrumentationRequest(it: Message<JsonObject>) {
        try {
            val command = LiveInstrumentCommand(it.body())
            when (command.commandType) {
                CommandType.ADD_LIVE_INSTRUMENT -> addInstrument(command)
                CommandType.REMOVE_LIVE_INSTRUMENT -> removeInstrument(command)
            }
        } catch (ex: InvocationTargetException) {
            if (ex.cause != null) {
                publishCommandError(it, ex.cause!!)
            } else {
                publishCommandError(it, ex.targetException)
            }
        } catch (ex: Throwable) {
            publishCommandError(it, ex)
        }
    }

    private fun publishCommandError(it: Message<JsonObject>, ex: Throwable) {
        val map: MutableMap<String, Any> = HashMap()
        map["command"] = it.body().toString()
        map["occurredAt"] = System.currentTimeMillis()
        map["cause"] = ThrowableTransformer.INSTANCE.convert2String(ex, 4000)
        log.error("Error occurred while processing command", ex)

        FrameHelper.sendFrame(
            BridgeEventType.PUBLISH.name.lowercase(),
            ProcessorAddress.LIVE_INSTRUMENT_REMOVED,
            null,
            SourceProbe.probeMessageHeaders,
            false,
            JsonObject.mapFrom(map),
            SourceProbe.tcpSocket
        )
    }

    private fun addInstrument(command: LiveInstrumentCommand) {
        if (ProbeConfiguration.isNotQuite) println("Adding instrument: $command")
        applyInstrument!!.invoke(null, command.instruments.first()) //todo: check for multiple
    }

    private fun removeInstrument(command: LiveInstrumentCommand) {
        for (breakpoint in command.instruments) {
            val breakpointId = breakpoint.id
            val location = breakpoint.location
            removeInstrument!!.invoke(null, location.source, location.line, breakpointId)
        }
        for (location in command.locations) {
            removeInstrument!!.invoke(null, location.source, location.line, null)
        }
    }

    companion object {
        private val log = LogManager.getLogger(LiveInstrumentRemote::class.java)
        private val EVENT_CONSUMER = BiConsumer(fun(address: String?, json: String?) {
            if (ProbeConfiguration.isNotQuite) println("Publishing event: $address, $json")
            FrameHelper.sendFrame(
                BridgeEventType.PUBLISH.name.lowercase(),
                address,
                null,
                SourceProbe.probeMessageHeaders,
                false,
                JsonObject(json),
                SourceProbe.tcpSocket
            )
        })
        private var putBreakpoint: Method? = null
        private var putLog: Method? = null
        private var putMeter: Method? = null
        private var openLocalSpan: Method? = null
        private var closeLocalSpan: Method? = null
        private var isInstrumentEnabled: Method? = null
        private var putLocalVariable: Method? = null
        private var putField: Method? = null
        private var putStaticField: Method? = null
        private var isHit: Method? = null

        @JvmStatic
        fun isInstrumentEnabled(instrumentId: String): Boolean {
            return try {
                isInstrumentEnabled!!.invoke(null, instrumentId) as Boolean
            } catch (e: Exception) {
                e.printStackTrace()
                false
            }
        }

        @JvmStatic
        fun isHit(breakpointId: String): Boolean {
            return try {
                isHit!!.invoke(null, breakpointId) as Boolean
            } catch (e: Exception) {
                e.printStackTrace()
                false
            }
        }

        @JvmStatic
        fun putBreakpoint(breakpointId: String, source: String, line: Int, ex: Throwable?) {
            try {
                putBreakpoint!!.invoke(null, breakpointId, source, line, ex)
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }

        @JvmStatic
        fun putLog(logId: String, logFormat: String, vararg logArguments: String?) {
            try {
                putLog!!.invoke(null, logId, logFormat, logArguments)
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }

        @JvmStatic
        fun putMeter(meterId: String) {
            try {
                putMeter!!.invoke(null, meterId)
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }

        @JvmStatic
        fun openLocalSpan(spanId: String) {
            try {
                openLocalSpan!!.invoke(null, spanId)
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }

        @JvmStatic
        fun closeLocalSpan(spanId: String) {
            try {
                closeLocalSpan!!.invoke(null, spanId, null)
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }

        @JvmStatic
        fun closeLocalSpanAndThrowException(throwable: Throwable, spanId: String): Throwable {
            try {
                closeLocalSpan!!.invoke(null, spanId, throwable)
            } catch (e: Exception) {
                e.printStackTrace()
            }
            throw throwable
        }

        @JvmStatic
        fun putLocalVariable(breakpointId: String, key: String, value: Any?, type: String) {
            try {
                putLocalVariable!!.invoke(null, breakpointId, key, value, type)
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }

        @JvmStatic
        fun putField(breakpointId: String, key: String, value: Any?, type: String?) {
            try {
                putField!!.invoke(null, breakpointId, key, value, type)
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }

        @JvmStatic
        fun putStaticField(breakpointId: String, key: String, value: Any?, type: String) {
            try {
                putStaticField!!.invoke(null, breakpointId, key, value, type)
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }
}
